LibC changes and useradd utility and documentation (#1059)
SerenityOS/serenity: Base/usr/share/man/man8/useradd.md

## Name

useradd - add a new user to the system password file

## Synopsis

```**sh
# useradd [options] LOGIN
```

## Description

This program uses the interface defined `pwd.h` to add a new user to the system

> password file. 

Reviewer: This implementation detail is hardly the most important thing about describing `useradd` :)

We can just say that it adds a new user to the system.

A user can either be a `login` user (that is, they have a home
directory; someone who actually uses to system) or a `system` user (such

> as the `lookup` user found in `/etc/passwd`). 

Reviewer: Is any of this relevant to the `useradd` utility? I don't have to specify whether the account I'm creating is a `login` or `system` user, so why are we mentioning it at all? This text seems more suited for a man page describing how user accounts work in general.

Author: I suppose not. I'll remove it :)

By default, the user will be added to the `users` group, which has a gid

> of 100.

Reviewer: No need to mention the gid here.

This program must be run as root.

## Options

> * `-u` _UID_: The User Idenfitier for the new user. If this is not specified, a value starting from `1000` will be specified.

Reviewer: ```suggestion
* `-u` _UID_: The user identifier for the new user. If not specified, an unused UID above `1000` will be auto-generated.
```

Author: No capilisation even though it's an acronym?

> * `-g` _GID_: The Group Identifier for the new user. If this is not specified, it will default to 100 (the Serenity **user** group).

Reviewer: ```suggestion
* `-g` _GID_: The group identifier for the new user. If not specified, it will default to 100 (the **users** group).
```

* `-s` _SHELLBIN_: The shell binary for this login. The default is `/bin/sh`. 

> * `-m`: Create a home directory for this new user at `/home/username>`.

Reviewer: ```suggestion
* `-m`: Create a home directory for this new user at `/home/username`.
```

* `-d` _HOMEPATH_: Set the home directory for this user to _HOMEPATH).

> * `-n` _GECOS_: GECOS information about this login. See [here](https://en.wikipedia.org/wiki/Gecos_field) for more information.

Reviewer: ```suggestion
* `-n` _GECOS_: GECOS information about this login. See [Wikipedia](https://en.wikipedia.org/wiki/Gecos_field) for more information.
```

## Examples

```sh

> # su

Reviewer: No need to show `su` here. Also the `#` indicates that we're already `root`. :)

Author: Ahh I see!

## Name

useradd - add a new user to the system password file

## Synopsis

```**sh
# useradd [options] LOGIN
```

## Description

This program uses adds a new user to the system.

By default, the user will be added to the `users` group.

This program must be run as root.

## Options

* `-u` _UID_: The user identifier for the new user. If not specified, an unused UID above `1000` will be auto-generated.
* `-g` _GID_: The group identifier for the new user. If not specified, it will default to 100 (the **users** group).
* `-s` _SHELLBIN_: The shell binary for this login. The default is `/bin/sh`. 
* `-m`: Create a home directory for this new user at `/home/username`.
* `-d` _HOMEPATH_: Set the home directory for this user to _HOMEPATH).
* `-n` _GECOS_: GECOS information about this login. See [Wikipedia](https://en.wikipedia.org/wiki/Gecos_field) for more information.

## Examples

```sh
# useradd -u 300 -m kling
# useradd -m -u 400 -g 200 -n "Sergey Bugaev" bugaevc
# useradd quaker
# useradd -g 1000 -d /tmp/somedir danboid
# cat /etc/passwd
# ls -la /home
```

## Exit Values

0 - Success
1 - Couldn't update the password file
3 - Invalid argument to option
4 - UID already in use

> 12 - Couldn't create home directory

Reviewer 2: This should be a Markdown list (use `*`s or `-`s)

And this section should be between Options and Examples

Author: > And this section should be between Options and Examples

Sorry I've never written a manpage before, so I'm following the layout of the Linux ones.

Reviewer 2: Sure, follow what other systems do, but if they put exit status after examples, they're wrong :)

See Linux [`man-pages`(7)](https://linux.die.net/man/7/man-pages) for some guidance (but ignore most of the groff stuff, we use Markdown)

> # ls -la /home

Reviewer 2: I get what you mean by including this ("you might want to examine `/etc/passwd` and `/home` afterwards"), but the examples shouldn't really show you how to use unrelated commands.

Instead, let's have a Files section that describes how this command affects `/etc/passwd` and `/home`. [Example here](https://github.com/SerenityOS/serenity/blob/master/Base/usr/share/man/man8/mount.md#files).

## Name

useradd - add a new user to the system password file

## Synopsis

```**sh
# useradd [options] LOGIN
```

## Description

This program uses adds a new user to the system.

> By default, the user will be added to the `users` group.

Reviewer 2: That is GID 100, right? I'd mention the numeric value in the man page too.

Author: Yeah I think so, I'll have to double check.

This program must be run as root.

## Options

* `-u` _UID_: The user identifier for the new user. If not specified, an unused UID above `1000` will be auto-generated.

> * `-g` _GID_: The group identifier for the new user. If not specified, it will default to 100 (the **users** group).

Reviewer 2: Above, "users" is monospace, and here it's bold. Pick one :)

Author: bold it is!

> * `-s` _SHELLBIN_: The shell binary for this login. The default is `/bin/sh`. 

Reviewer 2: These options should list their long names too, and value names should be the same ones as in the code.

* `-m`: Create a home directory for this new user at `/home/username`.
* `-d` _HOMEPATH_: Set the home directory for this user to _HOMEPATH).
* `-n` _GECOS_: GECOS information about this login. See [Wikipedia](https://en.wikipedia.org/wiki/Gecos_field) for more information.

## Exit Values

* 0 - Success
* 1 - Couldn't update the password file
* 3 - Invalid argument to option
* 4 - UID already in use
* 12 - Couldn't create home directory

## Files

> * `/etc/passwd` - user account information (such as `UID` and `GID` ) is stored here.

Reviewer 2: Extra space 😜

It's not just that the user info is stored here, it's that `useradd` *appends* new record to this file.

Author: > appends

Ahh, that's the word I was looking for (at 1am no doubt :laughing:)

> * `/etc/passwd` - new user information (such as `UID` and `GID`) is appended to this file.

Reviewer 2: Here and above: "UID" and "GID" are not code/identifiers, no need to make them monospace or otherwise emphasize

* `/home/` - user home directroy is created here if the `-m` flag is specified.

## Examples

```sh
# useradd -u 300 -m kling
# useradd -m -u 400 -g 200 -n "Sergey Bugaev" bugaevc
# useradd quaker

> # useradd -g 1000 -d /tmp/somedir danboid

Reviewer 2: Now that we have long options, let's show some examples of using them, too!

## Name

useradd - add a new user to the system password file

## Synopsis

```**sh

> # useradd [options] LOGIN

Reviewer 2: It should be `<login>`, not `LOGIN`

```

## Description

This program uses adds a new user to the system.

By default, the user will be added to the **users** group (which has a GID of 100).

This program must be run as root.

## Options

* `-u`, `--uid` _uid_: The user identifier for the new user. If not specified, an unused UID above `1000` will be auto-generated.
* `-g`, `--gid` _gid_: The group identifier for the new user. If not specified, it will default to 100 (the **users** group).

> * `-s`, `--shell` _shell_: The shell binary for this login. The default is `/bin/Shell`. 

Reviewer 2: The value name should be `path-to-shell` :slightly_smiling_face:

> * `-m`, `--create-home`: Create a home directory for this new user at `/home/username`.

Reviewer 2: It's created at whatever path you specify with `--home-dir` (and with `/home/username` as fallback). And you need to indicate in some way that `username` is not literally `username`, but whatever username is specified.

Reviewer 2: And it's actually called `login`, not `username`, above?

> * `-d`, `--home-dir` _path_: Set the home directory for this user to path).

Reviewer 2: stray `)`

And missing "By default, ..." (this is where it should say about `/home/username`, not above)

Author: `Set the home directory for this user to path). By default, this is `/home/username`, where `username` is the value of `login`.`??

> * `-d`, `--home-dir` _path_: Set the home directory for this user to path). By default, this is `/home/username`, where `username` is the value of login.

Reviewer 2: Still the stray `)` though :slightly_frowning_face:

Author: REEEEE

* `-n`, `--gecos` _general-info_: GECOS information about this login. See [Wikipedia](https://en.wikipedia.org/wiki/Gecos_field) for more information.

## Exit Values

* 0 - Success
* 1 - Couldn't update the password file
* 3 - Invalid argument to option
* 4 - UID already in use
* 12 - Couldn't create home directory

## Files

* `/etc/passwd` - new user information (such as UID and GID) is appended to this file.
* `/home/` - user home directroy is created here if the `-m` flag is specified.

## Examples

```sh
# useradd -u 300 -m kling

> # useradd -m -u 400 --gecos 200 -n "Sergey Bugaev" bugaevc

Reviewer 2: Wait, does `-n` do something different from `--gecos`? I though it's the same option.

Author: whoops that's --gid

# useradd quaker

> # useradd --gid 1000 -d /tmp/somedir -n "Dan Boid" danboid

Reviewer 2: Is @danboid's last name "Boid"? I though it's MacDonald...

Author: HAHA cheers for that one, I couldn't think of it on the spot

Reviewer 3: Kling is convinced I am Mr Boid. You can call me Mr Boid too if you want?

Author: @danboid That _was_ the reference because I forgot your surname, but it's a `GECOS` field so it has to be correct ;)
